src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,157 @@
-package core.basesyntax;
-
-import java.util.List;
-
-public class MyLinkedList<T> implements MyLinkedListInterface<T> {
-    @Override
-    public void add(T value) {
-    }
-
-    @Override
-    public void add(T value, int index) {
-    }
-
-    @Override
-    public void addAll(List<T> list) {
-    }
-
-    @Override
-    public T get(int index) {
-        return null;
-    }
-
-    @Override
-    public T set(T value, int index) {
-        return null;
-    }
-
-    @Override
-    public T remove(int index) {
-        return null;
-    }
-
-    @Override
-    public boolean remove(T object) {
-        return false;
-    }
-
-    @Override
-    public int size() {
-        return 0;
-    }
-
-    @Override
-    public boolean isEmpty() {
-        return false;
-    }
-}
+package core.basesyntax;
+
+import java.util.List;
+
+public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private int size;
+    private Node<T> head;
+    private Node<T> tail;
+    private Node<T> current;
+
+    @Override
+    public void add(T value) {
+        if (size == 0) {
+            firstElement(value);
+        } else {
+            Node<T> temp = new Node<T>(tail, value, null);
+            tail.next = temp;
+            tail = temp;
+            size++;
+        }
+    }
+
+    @Override
+    public void add(T value, int index) {
+        if (index == size) {
+            add(value);
+        } else {
+            checkIndex(index);
+            Node<T> temp = getNodeByIndex(index);
+            Node<T> tempPrev = temp.prev;
+            Node<T> current = new Node<T>(tempPrev, value, temp);
+            temp.prev = current;
+            if (index == 0) {
+                head = current;
+            }
+            if (tempPrev != null) {
+                tempPrev.next = current;
+            }
+            size++;
+        }
+    }
+
+    public Node<T> getNodeByIndex(int index) {
+        Node<T> temp = head;
+        int counter = 0;
+        while (counter < index) {

```

Comment:
[Don’t iterate your nodes from the head to tail or vice versa.](https://mate-academy.github.io/jv-program-common-mistakes/java-core/collections/linked-list.html#dont-iterate-your-nodes-from-the-head-to-tail-or-vice-versa)